src/m1_facilitator/lib/TransactionExecutor.ts

```diff
@@ -12,17 +12,155 @@
+  private readonly gasPrice: BigNumber;
+
+  /** Avatar account object. */
+  private readonly avatarAccount: AvatarAccount;
+
+  /** Gas limit at which transaction was sent. */
+  private gas?: BigNumber;
+
+  /** Interval in ms at which Transaction record will be dequeued in regular interval.  */
+  private readonly pollInterval: number;
+
+  /** Set interval handle. */
+  private setIntervalHandle: NodeJS.Timer | null;
+
+  /**
+   * Constructor.
+   *
+   * @param transactionRepository Transaction repository instance.
+   * @param web3 The web3 provider instance to be used for fetching nonce.
+   * @param gasPrice Gas price at which transaction needs to be sent.
+   * @param avatarAccount Avatar account object.
+   */
+  public constructor(
+    transactionRepository: TransactionRepository,
+    web3: Web3,
+    gasPrice: BigNumber,
+    avatarAccount: AvatarAccount,
+  ) {
+    this.transactionRepository = transactionRepository;
+    this.web3 = web3;
+    this.gasPrice = gasPrice;
+    this.avatarAccount = avatarAccount;
+    this.pollInterval = 1 * 60 * 1000; // in ms
+    this.setIntervalHandle = null;
+  }
+
+  /**
+   * This method enqueues transactions to queue. TransactionRepository acts as
+   * queue here. The method is called by services.
+   *
+   * @param toAddress Contract address at which transaction needs to be sent.
+   * @param rawTx Raw transaction object.
+   */
+  public async add(toAddress: string, rawTx: TransactionObject<string>): Promise<void> {
+    const transaction = new Transaction(
+      this.avatarAccount.address,
+      toAddress,
+      rawTx.encodeABI(),
+      this.gasPrice,
+    );
+    await this.transactionRepository.save(transaction);
+  }
+
+  /**
+   * It starts processing pending transactions in regular intervals.
+   */
+  public async start(): Promise<void> {
+    await this.execute();
+    this.setIntervalHandle = setInterval(
+      async (): Promise<void> => this.execute(),
+      this.pollInterval,
+    );
+  }
+
+  /**
+   * Stops processing of transactions by clearing setIntervalHandle.
+   * It should be called when facilitator stops.
+   */
+  public async stop(): Promise<void> {
+    if (this.setIntervalHandle) {
+      clearInterval(this.setIntervalHandle);
+    }
+  }
+
+  /**
+   * Execute method does below:
+   * - Dequeues transaction record in regular interval
+   * - Process the transaction and constructs data for transaction to be executed
+   * - Sends the transaction by calling sendTransaction method.
+   * - Updates Transaction repository with information like transactionHash, gas, nonce
+   */
+  private async execute(): Promise<void> {
+    const transaction = await this.transactionRepository.dequeue();
+    const nonce = await this.avatarAccount.getNonce(this.web3);
+    if (transaction !== null) {
+      const txHash = await this.sendTransaction(transaction, nonce);
+      transaction.transactionHash = txHash;
+      transaction.gas = this.gas;
```

Comment:
Where is `this.gas` assigned?

Reply:
in sendTransaction method. After estimating gas we record estimated gas value to `this.gas` variable.